src/NServiceBus.Core/Transports/DeliveryConstraintCollection.cs

```diff
@@ -0,0 +1,44 @@
+namespace NServiceBus.Transports
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using NServiceBus.DeliveryConstraints;
+
+    /// <summary>
+    /// A collection of delivery constraints.
+    /// </summary>
+    public class DeliveryConstraintCollection
+    {
+        List<DeliveryConstraint> constraints;
+
+        internal DeliveryConstraintCollection(IEnumerable<DeliveryConstraint> constraints)
+        {
+            this.constraints = constraints.ToList();
+        }
+
+        /// <summary>
+        /// Tries to get and remove constraint of a given type.
+        /// </summary>
+        /// <typeparam name="T">Requested type.</typeparam>
+        /// <param name="constraint">Returned constraint or null.</param>
+        /// <returns>Success or failure</returns>
+        public bool TryRemove<T>(out T constraint) where T : DeliveryConstraint
+        {
+            constraint = constraints.OfType<T>().FirstOrDefault();
```

Comment:
`FirstOrDefault` says to the future person  "if there is more than one we will ignore it".
`SingleOrDefault` says to the future person "there will only ever be zero or one"

This means  when this code evolves forward in the future you can modify this method based on those assumption